trainer.py

```diff
@@ -92,12 +101,52 @@ def get_epochs(self):
     def get_test_data(self):
         return self._x_test, self._y_test
 
+    def get_prune_threshold(self):
+        return self._prune_threshold
+
     def add_checkpoint_callback(self, callback):
         self._checkpoint_callbacks.append(callback)
 
-    def go(self):
+    def _train(self):
         raise NotImplementedError
 
+    def _prune(self, model):
+        """ Prunes model (in place) using prune_threshold """
+        BaseTrainer.prune_trained_model(model, self._prune_threshold)
+
+    @staticmethod
+    def prune_trained_model(model, threshold):
+        if math.isnan(threshold):
+            return
+        # Layer 0 has no input edges, start from layer 1
+        pruned_edges = 0
+        for i in range(1, len(model.layers)):
+            weights = model.layers[i].get_weights()
+            input_weights = weights[0]  # weights[1] is the list of node biases
+            # The incoming edge weights of node N is incoming_edge_weights[N]
+            new_weights = []
+            for node_input_weights in input_weights:
```

Comment:
I think this would run faster if it will be implemented as numpy matrix operations:

`w = np.array(weights)`
`new_weights = np.where(w<threshold, 0, w)`

or something like that

Reply:
Looks good, but how would I compute `pruned_edges` though in a numpy way? ATM I'm just iterating

Reply:
Use [numpy.count_nonzero](https://docs.scipy.org/doc/numpy/reference/generated/numpy.count_nonzero.html) :)

Reply:
I'm having a problem using your suggestion... `np.where(w<threshold, 0, w)` isn't good enough because if `w` is negative with large absolute value it will still be nullified. But `np.where(math.fabs(w)<threshold, 0, w)` complains when I try to do `math.fabs(w)` where `w` is not a 1-dimensional array.
I can do:
```
w = np.array(weights)
w_abs = np.absolute(w)
new_weights = np.where(w_abs<threshold, 0, w_abs)
```
...but the `np.absolute(w)` call seems expensive. Do you think it's worth it?

Reply:
nvm the answer is `(w<threshold) & (w>-threshold)`